### PR TITLE
Parse CSVs with newline

### PR DESCRIPTION
**Pull request by andrew-wm-arthur**
I replaced the CSV parser with an adaptation of the builtin parser in encoding/csv.

### CODE REVIEW

File: go/libraries/doltcore/table/untyped/csv/line_test.go [hunk 1]
```diff
@@ -48,32 +73,25 @@ func TestCSVSplitLine(t *testing.T) {
 		escapeQuotes   bool
 		expectErr      bool
 	}{
-		{", ,\t,\r,\n, \r\n\t", ",", []*string{nil, nil, nil, nil, nil, nil}, true, false},
-		{"\"\",\" \",\"\t\",\"\r\",\"\n\",\" \r\n\t\"", ",", toPointer([]string{"", " ", "\t", "\r", "\n", " \r\n\t"}), true, false},
+		{", ,\t,\t ,\t, \t", ",", []*string{nil, nil, nil, nil, nil, nil}, true, false},
```
Comment (andrew-wm-arthur) [inline go/libraries/doltcore/table/untyped/csv/line_test.go]: newlines must be quoted

File: go/libraries/doltcore/table/untyped/csv/line_test.go [hunk 2]
```diff
@@ -48,32 +73,25 @@ func TestCSVSplitLine(t *testing.T) {
...
+		{"\"\",\" \",\"\t\",\"\r\",\"\n\",\" \r\n\t\"", ",", toPointer([]string{"", " ", "\t", "\r", "\n", " \n\t"}), true, false},
```
Comment (andrew-wm-arthur) [inline go/libraries/doltcore/table/untyped/csv/line_test.go]: Any `\r\n` sequence in the input is replaced by `\n`

File: go/libraries/doltcore/table/untyped/csv/line_test.go [hunk 3]
```diff
@@ -48,32 +73,25 @@ func TestCSVSplitLine(t *testing.T) {
...
 		{`"", one, ""`, ",", toPointer([]string{``, `one`, ``}), true, false},
 		{`"", "", ""`, ",", toPointer([]string{``, ``, ``}), true, false},
-		{`"", one, ""`, ",", toPointer([]string{`""`, `one`, `""`}), false, false},
```
Comment (andrew-wm-arthur) [inline go/libraries/doltcore/table/untyped/csv/line_test.go]: These removed tests were parsing unescaped quotes

File: go/libraries/doltcore/table/untyped/csv/reader.go [hunk 4]
```diff
@@ -39,10 +43,25 @@ var ReadBufSize = 256 * 1024
...
 	closer io.Closer
 	bRd    *bufio.Reader
-	info   *CSVFileInfo
 	sch    schema.Schema
 	isDone bool
 	nbf    *types.NomsBinFormat
+
+	// CSV parsing is based on the standard Golang csv parser in encoding/csv/reader.go
+	// This parser has been adapted to differentiate between quoted and unquoted
+	// empty strings, and to use multi-rune delimiters. This adaptation removes the
```
Comment (timsehn) [inline go/libraries/doltcore/table/untyped/csv/reader.go]: "Multi-rune" made me laugh

File: go/libraries/doltcore/table/untyped/csv/line_test.go [hunk 5]
```diff
@@ -48,32 +73,25 @@ func TestCSVSplitLine(t *testing.T) {
 		escapeQuotes   bool
```
Comment (bheni) [inline go/libraries/doltcore/table/untyped/csv/line_test.go]: You should remove this parameter from the structure.

File: go/libraries/doltcore/table/untyped/csv/reader.go [hunk 6]
```diff
@@ -153,32 +204,200 @@ func (csvr *CSVReader) Close(ctx context.Context) error {
...
-		return nil, table.NewBadRow(nil, err.Error())
+func validDelim(s string) bool {
+	return !(strings.Contains(s, "\"") ||
+		strings.Contains(s, "\r") ||
+		strings.Contains(s, "\n") ||
+		strings.Contains(s, string(0xFFFD))) // Unicode replacement char
+}
+
+// nextRune returns the next rune in b or utf8.RuneError.
+func nextRune(b []byte) rune {
```
Comment (bheni) [inline go/libraries/doltcore/table/untyped/csv/reader.go]: Not sure this function is necessary.

File: go/libraries/doltcore/table/untyped/csv/reader.go [hunk 7]
```diff
@@ -153,32 +204,200 @@ func (csvr *CSVReader) Close(ctx context.Context) error {
...
-	sch := csvr.sch
-	allCols := sch.GetAllCols()
-	numCols := allCols.Size()
-	if len(colVals) != numCols {
-		return nil, table.NewBadRow(nil,
-			fmt.Sprintf("csv reader's schema expects %d fields, but line only has %d values.", numCols, len(colVals)),
-			fmt.Sprintf("line: '%s'", line),
-		)
+// readLine reads the next line (with the trailing endline).
+// If EOF is hit without a trailing endline, it will be omitted.
```
Comment (bheni) [inline go/libraries/doltcore/table/untyped/csv/reader.go]: What will be omitted?  The line will be omitted?

File: go/libraries/doltcore/table/untyped/csv/reader.go [hunk 8]
```diff
@@ -39,10 +43,25 @@ var ReadBufSize = 256 * 1024
...
 	nbf    *types.NomsBinFormat
+
+	// CSV parsing is based on the standard Golang csv parser in encoding/csv/reader.go
+	// This parser has been adapted to differentiate between quoted and unquoted
+	// empty strings, and to use multi-rune delimiters. This adaptation removes the
+	// comment feature and the lazyQuotes option
+	delim   []byte
+	numLine int
+	// rawBuffer is a line buffer only used by the readLine method.
+	rawBuffer []byte
```
Comment (bheni) [inline go/libraries/doltcore/table/untyped/csv/reader.go]: why is this a member of the struct and not just a local inside readLine?  I'd be surprised if reusing this buffer was a big performance benefit... but maybe on csvs that have really wide rows.

File: go/libraries/doltcore/table/untyped/csv/reader.go [hunk 9]
```diff
@@ -39,10 +43,25 @@ var ReadBufSize = 256 * 1024
...
+	// recordBuffer holds the unescaped fields, one after another.
+	// The fields can be accessed by using the indexes in fieldIndexes.
+	// E.g., For the row `a,"b","c""d",e`, recordBuffer will contain `abc"de`
+	// and fieldIndexes will contain the indexes [1, 2, 5, 6].
+	recordBuffer    []byte
+	fieldIndexes    []int
```
Comment (bheni) [inline go/libraries/doltcore/table/untyped/csv/reader.go]: Why are these members of the struct and not just locals in csvReadRecords.

File: go/libraries/doltcore/table/untyped/csv/reader.go [hunk 10]
```diff
@@ -153,32 +204,200 @@ func (csvr *CSVReader) Close(ctx context.Context) error {
...
+	return line, err
+}
 
-	taggedVals := make(row.TaggedValues)
-	for i := 0; i < allCols.Size(); i++ {
-		col := allCols.GetByIndex(i)
-		if colVals[i] == nil {
-			taggedVals[col.Tag] = nil
-			continue
+func (csvr *CSVReader) csvReadRecords(dst []*string) ([]*string, error) {
```
Comment (bheni) [inline go/libraries/doltcore/table/untyped/csv/reader.go]: this function is a monster. Would be nice if you could break some of the chunks up into functions with self documenting names.

File: go/libraries/doltcore/table/untyped/csv/reader.go [hunk 11]
```diff
@@ -153,32 +204,200 @@ func (csvr *CSVReader) Close(ctx context.Context) error {
...
+			}
+			csvr.recordBuffer = append(csvr.recordBuffer, field...)
+			csvr.fieldIndexes = append(csvr.fieldIndexes, len(csvr.recordBuffer))
+			// discard unquoted empty strings
+			keepString = append(keepString, len(field) != 0)
+			if i >= 0 {
+				line = line[i+delimLen:]
+				continue parseField
+			}
+			break parseField
```
Comment (bheni) [inline go/libraries/doltcore/table/untyped/csv/reader.go]: having to use labels is a likely sign your function has gotten too big :)

File: go/libraries/doltcore/table/untyped/csv/reader.go [hunk 12]
```diff
@@ -153,32 +204,200 @@ func (csvr *CSVReader) Close(ctx context.Context) error {
...
+					csvr.recordBuffer = append(csvr.recordBuffer, line[:i]...)
+					line = line[i+quoteLen:]
+					if atDelim(line, csvr.delim) {
+						// `"<delimiter>` sequence (end of field).
+						line = line[delimLen:]
+						csvr.fieldIndexes = append(csvr.fieldIndexes, len(csvr.recordBuffer))
+						keepString = append(keepString, true)
+						continue parseField
+					} else {
+						switch rn := nextRune(line); {
```
Comment (bheni) [inline go/libraries/doltcore/table/untyped/csv/reader.go]: I'm not sure about this switch statement.  Switches should either be switched on something:

    switch x {
        case 1:
        case 2:
        ...
    }

or should be switched without an expression

    switch {
        case x == 1:
        case x == 2:
        ...
    }

now you can switch on something you initialize in the switch:

    switch x := f(); x {
        case 1:
        case 2:
    }

and I already think that starts to look confusing conflating the initialization with the value you are switching on, but using the semicolon to initialize a value and then switching on an empty expression in order to save 1 line of code isn't worth the confusion.

    rn := nextRune(line)
    switch {
        ...

is better.

**Pull request closed without merging**